rust-symcrypt/src/ecdh.rs

```diff
@@ -0,0 +1,270 @@
+/* EcDh functions. For further documentation please refer to symcrypt.h */
+
+use crate::curve_type::*;
+use crate::errors::SymCryptError;
+use std::vec;
+use symcrypt_sys;
+
+/// EcDhKey is a wrapper around symcrypt_sys::PSYMCRYPT_ECKEY, this is to let rust handle the free'ing of this pointer
+/// EcDhKey must be allocated after EcDhExpandedCurve and free'd before EcDhExpandedCurve is free'd
+struct EcDhKey {
+    inner: symcrypt_sys::PSYMCRYPT_ECKEY,
+    expanded_curve: EcDhExpandedCurve,
```

Comment:
Is it ever possible for multiple keys to share the same expanded curve?

Reply:
Yes, it is. The key does not own the curve; it is the responsibility of whoever called `SymCryptEcurveAllocate` to clean it up by calling `SymCryptEcurveFree`

Reply:
Does the curve object have a ref count internally? If so, we could reflect that here to do shared ownership of the keys.

If not, another option is to have the Rust curve wrapper use an `Arc` or `Rc` and add shared ownership on top of the underlying SymCrypt structures. That means we'd have an extra allocation and a layer of indirection though, so it may not be worthwhile.

Reply:
No, it’s not refcounted by SymCrypt.

Reply:
Khang is planning on using lazy static to call SymCryptEcurveAllocate(curve) on first use and share for subsequent calls.

Reply:
Should only need one instance per curve type per process.

I'm seeing in SymCrypt-OpenSSL it calls SymCryptEcurveAllocate() for each supported curve type during engine init.

We will be better in rust, where we defer until the curve is actually used.